src/bucket.ts

```diff
@@ -3174,6 +3189,177 @@ class Bucket extends ServiceObject {
+   * @property {string|error} status Status of the file upload.
+   *     "success" or an Error returned while attempting to upload current file.
+   */
+  /**
+   * @typedef {array} UploadDirectoryResponse
+   * @property {UploadFileResponse[]} 0 List of {@link Bucket#UploadFileResponse}
+   *     objects for each of the uploaded files.
+   */
+  /**
+   * @callback UploadDirectoryCallback
+   * @param {Error} err Request error, if any.
+   * @param {UploadFileResponse[]} resp List of {@link Bucket#UploadFileResponse}
+   *     objects for every uploaded file.
+   */
+
+  /**
+   * This is a convenience method for making multiple {@link Bucket#upload}
+   * calls.
+   *
+   * You can specify whether or not an upload is resumable by setting
+   * `options.resumable`. *Resumable uploads are enabled by default if your
+   * input file is larger than 5 MB.*
+   *
+   * For faster crc32c computation, you must manually install
+   * [`fast-crc32c`](http://www.gitnpm.com/fast-crc32c):
+   *
+   *     $ npm install --save fast-crc32c
+   *
+   * @see [Upload Options (Simple or Resumable)]{@link https://cloud.google.com/storage/docs/json_api/v1/how-tos/upload#uploads}
+   * @see [Objects: insert API Documentation]{@link https://cloud.google.com/storage/docs/json_api/v1/objects/insert}
+   *
+   * @param {string} directoryPathString The fully qualified path to the
+   *     directory you wish to upload to your bucket.
+   * @param {UploadDirectoryOptions} [options] Configuration options.
+   * @param {UploadDirectoryCallback} [callback] Callback function.
+   * @returns {Promise<UploadDirectoryResponse>}
+   *
+   * @example <caption>include:samples/files.js</caption>
+   * region_tag:storage_upload_top_level_files
+   * Another example:
+   *
+   * @example <caption>include:samples/files.js</caption>
+   * region_tag:storage_upload_full_folder_hierarchy
+   *
+   * //-
+   * // If the callback is omitted, we'll return a Promise.
+   * //-
+   * const [response] = await bucket.uploadDirectory('local-directory-path');
+   */
+  uploadDirectory(
+    directoryPathString: string,
+    optionsOrCallback?: UploadDirectoryOptions | UploadDirectoryCallback,
+    callback?: UploadDirectoryCallback
+  ): Promise<UploadDirectoryResponse> | void {
+    const options =
+      typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
+    callback =
+      typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
+
+    let stat: fs.Stats | Error;
+    try {
+      stat = fs.lstatSync(directoryPathString);
+    } catch (error) {
+      stat = error;
+    }
+    if (stat instanceof Error || !stat.isDirectory()) {
+      throw new Error(`${directoryPathString} is an invalid directory`);
+    }
+
+    const fileList: string[] = [];
+    const pathDirName = path.dirname(directoryPathString);
+
+    getFileList(directoryPathString);
+
+    function getFileList(directory: string) {
+      const items = fs.readdirSync(directory);
```

Comment:
I think we want to use the async versions of all the `fs` methods, I believe the sync versions are all blocking operations.